group_theory/perm.lean

```diff
@@ -286,9 +385,204 @@ funext $ λ f,
+        by rw [← h _ x.2 this]; exact mt (hi _ _ this x.2) x.2⟩ : {y // g y ≠ y})),
+      from ⟨λ ⟨x, hx⟩ ⟨y, hy⟩ h, subtype.eq (hi _ _ _ _ (subtype.mk.inj h)),
+        λ ⟨y, hy⟩, let ⟨x, hfx, hx⟩ := hg y hy in ⟨⟨x, hfx⟩, subtype.eq hx⟩⟩))
+    (λ ⟨x, _⟩, subtype.eq (h x _ _))
+... = sign g : sign_subtype_perm _ _ (λ _, id)
+
+lemma is_cycle_swap {x y : α} (hxy : x ≠ y) : is_cycle (swap x y) :=
+⟨y, by rwa swap_apply_right,
+λ a (ha : ite (a = x) y (ite (a = y) x a) ≠ a),
+if hya : y = a then ⟨0, hya⟩
+  else ⟨1, by rw [gpow_one, swap_apply_def]; split_ifs at *; cc⟩⟩
+
+lemma is_cycle_inv {f : perm α} (hf : is_cycle f) : is_cycle (f⁻¹) :=
+let ⟨x, hx⟩ := hf in
+⟨x, by simp [eq_inv_iff_eq, inv_eq_iff_eq, *] at *; cc,
+  λ y hy, let ⟨i, hi⟩ := hx.2 y (by simp [eq_inv_iff_eq, inv_eq_iff_eq, *] at *; cc) in
+  ⟨-i, by rwa [gpow_neg, inv_gpow, inv_inv]⟩⟩
+
+lemma swap_mul_eq_mul_swap (f : perm α) (x y : α) : swap x y * f = f * swap (f⁻¹ x) (f⁻¹ y) :=
+equiv.ext _ _ $ λ z, begin
+  simp [mul_apply, swap_apply_def],
+  split_ifs;
+  simp [*, eq_inv_iff_eq] at * <|> cc
+end
+
+lemma mul_swap_eq_swap_mul (f : perm α) (x y : α) : f * swap x y = swap (f x) (f y) * f :=
+by rw [swap_mul_eq_mul_swap, inv_apply_self, inv_apply_self]
+
+lemma is_cycle_swap_mul_aux : ∀ (n : ℤ) {b x : α} {f : perm α} (hf : is_cycle f)
+  (hb : (swap x (f x) * f) b ≠ b)
+  (h : (f ^ n) (f x) = b), ∃ i : ℤ, ((swap x (f x) * f) ^ i) (f x) = b
+| 0         := λ b x f hf hb h, ⟨0, h⟩
+| (n+1 : ℕ) := λ b x f hf hb h,
+  if hfbx : f x = b then ⟨0, hfbx⟩ else
+  have f b ≠ b ∧ b ≠ x := support_swap_mul hb,
+  have hb : (swap x (f x) * f) (f⁻¹ b) ≠ f⁻¹ b,
+    by rw [mul_apply, apply_inv_self, swap_apply_of_ne_of_ne this.2 (ne.symm hfbx),
+        ne.def, ← injective.eq_iff f.bijective.1, apply_inv_self];
+      exact this.1,
+  have wf : n < int.nat_abs (1 + ↑n),
+    by rw [add_comm, ← int.coe_nat_one, ← int.coe_nat_add, int.nat_abs_of_nat];
+    exact nat.lt_succ_self n,
+  let ⟨i, hi⟩ := is_cycle_swap_mul_aux n hf hb
+    (f.bijective.1 $ by
+      rw [apply_inv_self];
+      rwa [gpow_coe_nat, pow_succ, mul_apply] at h) in
+  ⟨i + 1, by rw [add_comm, gpow_add, mul_apply, hi, gpow_one, mul_apply, apply_inv_self,
+      swap_apply_of_ne_of_ne this.2 (ne.symm hfbx)]⟩
+| -[1+ n] := λ b x f hf hb h,
+  if hfbx : f⁻¹ x = b then ⟨-1, by rwa [gpow_neg, gpow_one, mul_inv_rev, mul_apply, swap_inv, swap_apply_right]⟩
+  else if hfbx' : f x = b then ⟨0, hfbx'⟩
+  else
+  have f b ≠ b ∧ b ≠ x := support_swap_mul hb,
+  have hb : (swap x (f⁻¹ x) * f⁻¹) (f⁻¹ b) ≠ f⁻¹ b,
+    by rw [mul_apply, swap_apply_def];
+      split_ifs;
+      simp [inv_eq_iff_eq, eq_inv_iff_eq] at *; cc,
+  have wf : n < n.succ, from nat.lt_succ_self n,
+  let ⟨i, hi⟩ := is_cycle_swap_mul_aux n (is_cycle_inv hf) hb
+    (show (f⁻¹ ^ ↑n) (f⁻¹ x) = f⁻¹ b, by
+      rw [← h, ← mul_apply, ← mul_apply, ← mul_apply, gpow_neg_succ, ← inv_pow, pow_succ', mul_assoc,
```

Comment:
`by` is one of the exceptions where it is put on the next line together with the tactic.